src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,747 @@
+     * threads invoking the returned function's
+     * {@linkplain IntFunction#apply(int) apply()} method when a value is already under
+     * computation will block until a value is computed or an exception is thrown by
+     * the computing thread.
+     * <p>
+     * If the provided {@code original} function throws an exception, it is relayed
+     * to the initial caller and no content is recorded.
+     * <p>
+     * If the provided {@code original} function recursively calls the returned
+     * function for the same index, an {@linkplain IllegalStateException} will
+     * be thrown.
+     *
+     * @param size     the size of the allowed inputs in {@code [0, size)}
+     * @param original IntFunction used to compute cached values
+     * @param <R>      the type of results delivered by the returned IntFunction
+     * @throws IllegalArgumentException if the provided {@code size} is negative.
+     */
+    static <R> IntFunction<R> intFunction(int size,
+                                          IntFunction<? extends R> original) {
+        if (size < 0) {
+            throw new IllegalArgumentException();
+        }
+        Objects.requireNonNull(original);
+        return StableIntFunction.of(size, original);
+    }
+
+    /**
+     * {@return a new stable {@linkplain Function}}
+     * <p>
+     * The returned function is a caching function that, for each allowed
+     * input in the given set of {@code inputs}, records the values of the provided
+     * {@code original} function upon being first accessed via the returned function's
+     * {@linkplain Function#apply(Object) apply()} method.
+     * <p>
+     * The provided {@code original} function is guaranteed to be successfully invoked
+     * at most once per allowed input, even in a multi-threaded environment. Competing
+     * threads invoking the returned function's {@linkplain Function#apply(Object) apply()}
+     * method when a value is already under computation will block until a value is
+     * computed or an exception is thrown by the computing thread.
+     * <p>
+     * If the provided {@code original} function throws an exception, it is relayed to
+     * the initial caller and no content is recorded.
+     * <p>
+     * If the provided {@code original} function recursively calls the returned
+     * function for the same input, an {@linkplain IllegalStateException} will
+     * be thrown.
+     *
+     * @param inputs   the set of (non-null) allowed input values
+     * @param original Function used to compute cached values
+     * @param <T>      the type of the input to the returned Function
+     * @param <R>      the type of results delivered by the returned Function
+     * @throws NullPointerException if the provided set of {@code inputs} contains a
+     *                              {@code null} element.
+     */
+    static <T, R> Function<T, R> function(Set<? extends T> inputs,
```

Comment:
The javadoc should say that the returned function throws IAE if the input is not supported, right? An alternative would be to have a factory that returns `Function<T, Optional<R>>` -- although that would be more cumbersome to use.

Reply:
And, same for `int function`